salt/modules/vsphere.py

```diff
@@ -9050,6 +9094,492 @@ def _delete_device(device):
     return device_spec
 
 
+@depends(HAS_PYVMOMI, HAS_VSPHERE_SDK)
+@supports_proxies('vcenter')
+@gets_service_instance_via_proxy
+def list_tag_categories(server=None, username=None, password=None,
+                        service_instance=None):
+    '''
+    List existing categories a user has access to.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+            salt vm_minion vsphere.list_tag_categories
+
+    :param basestring server:
+        Target DNS or IP of vCenter center.
+    :param basestring username:
+        Username associated with the vCenter center.
+    :param basestring password:
+        Password associated with the vCenter center.
+    :returns:
+        Value(s) of category_id.
+    :rtype:
+        list of str
+    '''
+    # Get salted vSphere Client
+    if not (server and username and password):
+        # User didn't provide CLI args so use proxy information
+        details = __salt__['vcenter.get_details']()
+        server = details['vcenter']
+        username = details['username']
+        password = details['password']
+
+    # Establish connection with client
+    client = salt.utils.vmware.get_vsphere_client(server=server,
+                                                  username=username,
+                                                  password=password)
```

Comment:
Getting a client instance should be factored in it's own private function, since its used more than once.

```python
def _get_client(server, username, password):
    return blah
```
or
```python
def _get_client(details):
    server = details['vcenter']
    .... etc ....
    return client_instance
```
No?

Reply:
Fixing that now...